src/cplscheme/BaseCouplingScheme.cpp

```diff
@@ -724,7 +724,20 @@ void BaseCouplingScheme::advanceTXTWriters()
 
 bool BaseCouplingScheme::reachedEndOfTimeWindow()
 {
-  return math::equals(getNextTimeStepMaxSize(), 0.0, _eps) || not hasTimeWindowSize();
+  if (math::equals(getNextTimeStepMaxSize(), 0.0, _minTimeStepSize) && getNextTimeStepMaxSize() > math::NUMERICAL_ZERO_DIFFERENCE && hasTimeWindowSize()) {
+    for (auto &data : _allData | boost::adaptors::map_values) {
+      auto   timeStepStorage = data->timeStepsStorage();
+      auto   values          = timeStepStorage.getTimesAndValues();
+      auto   ts              = values.first;
+      double endOfTimeWindow = getTime() - _computedTimeWindowPart + _timeWindowSize;
+      if (ts(ts.size() - 1) < endOfTimeWindow - math::NUMERICAL_ZERO_DIFFERENCE) {
```

Comment:
Instead of `ts(ts.size() - 1)` something like `lastT` would be much more readable. I thing you could even use `Storage::maxStoredTime()` here.